README.md
# WS2801-webserver

This is a ready-to-use webserver for the [WS2801-Pi package](https://www.npmjs.com/package/ws2801-pi).

There is also a [TypeScript client](https://www.npmjs.com/package/ws2801-client) for this webserver.

You may also want to take a look at [WS2801](https://github.com/SteffenKn/WS2801) which includes the [Webserver](https://www.npmjs.com/package/ws2801-webserver) (this package), the [Alexa handler](https://www.npmjs.com/package/ws2801-alexa), and the [LedController](https://www.npmjs.com/package/ws2801-pi).

## Wiring

The wiring of the Pi is described [here](https://github.com/SteffenKn/WS2801-PI#wiring).

## Usage

There are basically two ways to use WS2801-webserver:

### Using this repository

1. Clone this repository
2. Edit the config file as described [here](#Configuration)
3. Run the build script: `npm run build`
4. Run the start script: `npm start`

### Using the npm module

1. Create a new npm project folder
2. Install this module: `npm i ws2801-webserver`

```typescript
import {Config as WebserverConfig, Ws2801Webserver} from 'ws2801-webserver';

const config: WebserverConfig = {
  port: 45451,
  confirmationPort: 45452,
  amountOfLeds: 10,
  useAuth: true,
};

const webserver: Ws2801Webserver = new Ws2801Webserver(config);

webserver.start();
```

## Configuration

The config can be specified when initializing the Ws2801Webserver.
If no config was specified or if the repository is used, the defaultConfig which is stored [here](./src/config/config.ts).

The config can look like this:
```
{
  port: 45451,
  confirmationPort: 45452,
  amountOfLeds: 10,
  useAuth: true,
};
```

- `port` is used for the webserver that provides the most routes. This may be publicly accessible.
- `confirmationPort` is only needed if the auth mechanism is enabled (`useAuth` is set to true). This webserver provides the confirmation routes for new registrations. This should never be publicly accessible, at least until a proper confirmation website is available.
- `amountOfLeds` is used to specify how many leds are connected to the Pi. This is only necessary if no other ledController is specified in the constructor.
- `useAuth` is used to define whether the auth mechanism should be used or not.
## Functions

### constructor

#### Parameters:

- `config`
  - optional
  - Type: [Config](./src/types/config.ts)
  - The config for this module, as described [here](#Configuration). If no value is set, the [default config](./src/config/config.ts) is used.

- `ledController`
  - optional
  - Type: [LedController](https://github.com/SteffenKn/WS2801-PI/blob/develop/src/index.ts#L44)
  - The led controller that controls the leds of the connected led strip. If no value is specified, WS2801-webserver will create its own LedController with the LedAmount that is configured in the [default config](./src/config/config.ts).
    - To avoid rendering problems, there should be only one LedController instance.

### start

Starts the webserver.

### stop

Stops the webserver.

### getExpressServer

Returns the express server instance used by WS2801-webserver. This can be used to add more routes.

#### Returns

Returns the express server instance.
- Type: [express](https://www.npmjs.com/package/express)

## Routes

The routes marked with [Auth] can only be used if the auth mechanism is enabled.
If the auth mechanism is enabled, an apiKey must be provided as a query parameter for all routes except `/login-required` and `/register`. Otherwise a 403 is returned.

### GET: /login-required

Returns wheter the auth mechanism is enabled or not.

- Route: /login-required
- Method: Get
- Returns:
  - 200
    - Type: {loginRequired: boolean}

### Post: /register [Auth]

Registers a new user and waits for confirmation.

- Route: /register
- Method: Post
- Body: {name: string, apiKey: string}
- Returns:
  - 200
    - Registration was successful
      - Type: {apiKey: string}
  - 400
    - Body was incorrect
      - Type: string
  - 403
    - Username already exists
      - Type: string
    - Registration was not confirmed
      - Type: string
    - Registration was not successful
      - Type: string
### Post: /login [Auth]

Checks if the specified apiKey is working.

- Route: /login
- Method: Post
- Returns:
  - 200
    - Login was successful
      - Type: {apiKey: string}
  - 400
    - Body was incorrect
      - Type: string
  - 403
    - User in not registered
      - Type: string
    - User is not allowed to use the api
      - Type: string

### Get: /led-strip

Returns the current state of the led strip.

- Route: /led-strip
- Method: Get
- Returns:
  - 200
    - Type: {ledStrip: [LedStrip](https://github.com/SteffenKn/WS2801-PI/blob/develop/src/index.ts#L6)}

### Post: /led-strip/fill

Fills the led strip with a single color.

- Route /led-strip/fill
- Method: Post
- Body:
  - color
    - Type: [LedColor](https://github.com/SteffenKn/WS2801-PI/blob/develop/src/index.ts#L8-L12)
  - brightness
    - optional
    - Type: number
- Returns:
  - 200
    - The Led Strip was changed successfully.
      - Type: {ledStrip: [LedStrip](https://github.com/SteffenKn/WS2801-PI/blob/develop/src/index.ts#L6)}
  - 400
    - Body was incorrect
      - Type: string

### Post: /led-strip/clear

Clears the led strip.

- Route: /led-strip/clear
- Method: Post
- Returns:
  - 200
    - Type: {ledStrip: [LedStrip](https://github.com/SteffenKn/WS2801-PI/blob/develop/src/index.ts#L6)}

### Post: /led-strip/led/:ledIndex/set

- Route: /led-strip/led/:ledIndex/set
- Method: Post
- Route Parameter:
  - ledIndex: number
- Body:
  - color
    - Type: [LedColor](https://github.com/SteffenKn/WS2801-PI/blob/develop/src/index.ts#L8-L12)
  - brightness (brightness is set for the whole LED strip)
    - optional
    - Type: number
- Returns:
  - 200
    - The Led was successfully changed.
      - Type: {ledStrip: [LedStrip](https://github.com/SteffenKn/WS2801-PI/blob/develop/src/index.ts#L6)}
  - 400
    - Body was incorrect
      - Type: string

### Post: /led-strip/brightness/set

- Route: /led-strip/brightness/set
- Method: Post
- Body:
  - brightness (brightness is set for the whole LED strip)
    - Type: number
- Returns:
  - 200
    - The brightness successfully changed.
      - Type: string
  - 400
    - Body was incorrect
      - Type: string

### Get: /led-strip/brightness

- Route: /led-strip/brightness
- Method: Get
 - Returns: {brightness: number}

### Post: /led-strip/set

Sets all leds according to the provided colors.

- Route /led-strip/set
- Method: Post
- Body:
  - ledStrip
    - Type: [LedStrip](https://github.com/SteffenKn/WS2801-PI/blob/develop/src/index.ts#L6)
  - brightness
    - optional
    - Type: number
- Returns:
  - 200
    - The Led Strip was changed successfully.
      - Type: {ledStrip: [LedStrip](https://github.com/SteffenKn/WS2801-PI/blob/develop/src/index.ts#L6)}
  - 400
    - Body was incorrect
      - Type: string
    - Provided LedStrip was invalid
      - Type: string

### Post: /led-strip/animation/start

Executes the provided animation script. The animation script must be passed as a string.
The animation script can access the [LED controller](https://github.com/SteffenKn/WS2801-PI/blob/develop/src/index.ts#L44) via `ledController` and the amount of Leds via `ledAmount`.

- Route /led-strip/set
- Method: Post
- Body:
  - animationScript
    - Type: string
- Returns:
  - 200
    - The animation was started successfully
      - Type: string
  - 400
    - Body was incorrect
      - Type: string

### Delete: /led-strip/animation/stop

Cancels the currently running animation.

- Route /led-strip/set
- Method: Delete
- Returns:
  - 200
    - The animation was stopped successfully
      - Type: string

### Get: /led-strip/animation/finished

Returns as soon as the currently running animation finished.

- Route /led-strip/animation/finished
- Method: Get
- Returns:
  - 200
    - The animation was successfully finished.
      - Type: string


## Socket.Io Events

### LedStrip changed

Triggered every time the LedStrip is changed.

- Event: 'led-strip__changed'
- Payload: [LedStrip](https://github.com/SteffenKn/WS2801-PI/blob/develop/src/index.ts#L6)

### Brightness changed

Triggered ecery time the brightness gets changed.

- Event: 'brightness__changed'
- Payload: number

### Animation started

Triggered ecery time an animation was started.

- Event: 'animation__started'

### Animation finished

Triggered ecery time an animation was finished.

- Event: 'animation__finished'

### Animation stopped

Triggered ecery time an animation was stopped.

- Event: 'animation__stopped'

## Authorization

The auth mechanism provides simple authorization. It can be enabled or disabled via the config.
To check if a WS2801-webserver instance uses authorization, the [loginRequired route](#get-login-required) can be used.

The authorization works as follows:

- New user requests registration using the [register route](#post-register-auth), providing a username and an apiKey
- If the username does not already exist, a confirmation link gets logged in the console.
- Once the confirmation link is clicked, the user is registered.
- The user can then use the [login route](#post-login-auth) to verify that he is logged in.
- The api key must then be provided as a query parameter for all other api calls, otherwise a 403 error will be returned.

The user data is stored in `$(PWD)/.storage/webserver-api-keys.json`.


CHANGELOG.md
# Changelog

## v1

### v1.0.1

- **✨ Add logRequests to config**


src/app.ts
import {ChildProcess, fork} from 'child_process';
import express from 'express';
import path from 'path';

import Ws2801Pi, {LedColor, LedStrip} from 'ws2801-pi';

import {AuthService} from './auth-service';
import {Config} from './config/config';
import {validateLedStrip} from './led-strip-validation';
import {Webserver} from './webserver';

const webserver: Webserver = new Webserver(Config.port);
const authService: AuthService = new AuthService(webserver);

let currentAnimationProcess: ChildProcess;

const ledController: Ws2801Pi = new Ws2801Pi(Config.amountOfLeds);

async function run(): Promise<void> {
  webserver.start();

  if (Config.useAuth) {
    authService.start();
  }

  webserver.addGetRoute('/login-required', (_: express.Request, response: express.Response): void => {
    response.status(200).json({loginRequired: Config.useAuth});
  });

  webserver.addGetRoute('/led-strip', (_: express.Request, response: express.Response): void => {
    const ledStrip: LedStrip = ledController.getLedStrip();

    response.status(200).json({ledStrip: ledStrip});
  });

  webserver.addPostRoute('/led-strip/fill', async(request: express.Request, response: express.Response): Promise<void> => {
    const color: LedColor = request.body.color;

    if (!color) {
      response.status(400).send(`Request body must contain a 'color', e.g. {red: 255, green: 0, blue: 0}.`);

      return;
    }

    await ledController.fillLeds(color).show();

    const ledStrip: LedStrip = ledController.getLedStrip();

    response.status(200).json({ledStrip: ledStrip});
  });

  webserver.addPostRoute('/led-strip/clear', async(request: express.Request, response: express.Response): Promise<void> => {
    await ledController.clearLeds().show();

    const ledStrip: LedStrip = ledController.getLedStrip();

    response.status(200).json({ledStrip: ledStrip});
  });

  webserver.addPostRoute('/led-strip/led/:ledIndex/set', async(request: express.Request, response: express.Response): Promise<void> => {
    const enteredLedIndex: string = request.params.ledIndex;
    const ledIndex: number = parseInt(enteredLedIndex);

    const color: LedColor = request.body.color;

    if (Number.isNaN(ledIndex) || ledIndex < 0 || ledIndex >= Config.amountOfLeds) {
      response.status(400).send(`URL must contain a led index between 0 and ${Config.amountOfLeds - 1} (Received ${enteredLedIndex}).`);

      return;
    }

    if (!color) {
      response.status(400).send(`Request body must contain a 'color', e.g. {red: 255, green: 0, blue: 0}.`);

      return;
    }

    await ledController.setLed(ledIndex, color).show();

    const ledStrip: LedStrip = ledController.getLedStrip();

    response.status(200).json({ledStrip: ledStrip});
  });

  webserver.addPostRoute('/led-strip/brightness/set', async(request: express.Request, response: express.Response): Promise<void> => {
    const brightness: number | 'auto' = request.body.brightness;

    if (brightness == undefined || (typeof brightness !== 'number' && brightness !== 'auto') || brightness < 0 || brightness > 100) {
      response.status(400).send(`Request body must contain a 'brightness' (number between 0 and 100 or 'auto' for automatic brightness).`);

      return;
    }

    await ledController.setBrightness(brightness).show();

    if (currentAnimationProcess) {
      currentAnimationProcess.send({brightness: brightness});
    }

    const ledStrip: LedStrip = ledController.getLedStrip();

    response.status(200).json({ledStrip: ledStrip});
  });

  webserver.addGetRoute('/led-strip/brightness', async(request: express.Request, response: express.Response): Promise<void> => {
    const brightness: number | 'auto' = ledController.getBrightness();

    response.status(200).json({brightness: brightness});
  });

  webserver.addPostRoute('/led-strip/set', async(request: express.Request, response: express.Response): Promise<void> => {
    const ledStrip: LedStrip = request.body.ledStrip;

    if (!ledStrip) {
      response.status(400).send(`Request body must contain a 'ledStrip' (an array of LedColors with the length of the led strip).`);

      return;
    }

    try {
      validateLedStrip(Config.amountOfLeds, ledStrip);
    } catch (error) {
      response.status(400).send(error.message);

      return;
    }

    for (let ledIndex: number = 0; ledIndex < ledStrip.length; ledIndex++) {
      ledController.setLed(ledIndex, ledStrip[ledIndex]);
    }

    await ledController.show();

    const renderedLedStrip: LedStrip = ledController.getLedStrip();

    response.status(200).json({ledStrip: renderedLedStrip});
  });

  webserver.addPostRoute('/led-strip/animation/start', async(request: express.Request, response: express.Response): Promise<void> => {
    const animationScript: string = request.body.animationScript;

    if (animationScript == undefined) {
      response.status(400).send(`Request body must contain a 'animationScript'.`);

      return;
    }

    if (currentAnimationProcess) {
      currentAnimationProcess.kill();
      currentAnimationProcess = undefined;
    }

    const brightness: number | 'auto' = ledController.getBrightness();

    currentAnimationProcess = fork(path.join(__dirname, 'animator.js'), [animationScript, brightness.toString()], {});

    // tslint:disable-next-line: typedef no-any
    const eventCallback = (message: any): void => {
      currentAnimationProcess.removeListener('message', eventCallback);

      if (message === 'animation-finished') {
        currentAnimationProcess.kill();
        currentAnimationProcess = undefined;
      }
    };

    currentAnimationProcess.on('message', eventCallback);

    response.status(200).send('success!');
  });

  webserver.addDeleteRoute('/led-strip/animation/stop', async(request: express.Request, response: express.Response): Promise<void> => {
    if (currentAnimationProcess) {
      currentAnimationProcess.kill();
      currentAnimationProcess = undefined;
    }

    response.status(200).send('success!');
  });

  webserver.addGetRoute('/led-strip/animation/finished', async(request: express.Request, response: express.Response): Promise<void> => {
    // tslint:disable-next-line: typedef no-any
    const eventCallback = (message: any): void => {
      currentAnimationProcess.removeListener('message', eventCallback);

      if (message === 'animation-finished') {
        response.status(200).send('success!');
      }
    };

    currentAnimationProcess.on('message', eventCallback);
  });
}

run();
